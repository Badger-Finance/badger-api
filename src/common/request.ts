import axios from "axios";

export async function request<T>(baseURL: string, params?: Record<string, string>): Promise<T> {
  try {
    const { data } = await axios.get(baseURL, { params });
    return data as T;
  } catch (error) {
    throw error;
  }
}
