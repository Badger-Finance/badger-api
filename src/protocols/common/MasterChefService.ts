import { Service } from "@tsed/di";
import NodeCache from "node-cache";
import { MasterChefPool } from "../../interface/MasterChef";
import { Chain } from "../../util/constants";
import { getProvider } from "../../util/util";

@Service()
export class MasterChefService {
  private cache: NodeCache;

  constructor() {
    this.cache = new NodeCache({ stdTTL: 300, checkperiod: 480 });
  }

  async getPool(chain: Chain, chef: string, poolId: number): Promise<MasterChefPool> {
    const provider = getProvider(chain);
  }
}
