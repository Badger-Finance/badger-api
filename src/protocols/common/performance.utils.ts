import fetch from 'node-fetch';
import { Performance, uniformPerformance } from '../interfaces/performance.interface';
import { createValueSource, ValueSource } from '../interfaces/value-source.interface';

export async function getSwapValueSource(graphUrl: string, name: string, poolAddress: string): Promise<ValueSource> {
  // TODO: Move query to GraphService
  const query = `
    {
      pairDayDatas(first: 30, orderBy: date, orderDirection: desc, where:{pairAddress: "${poolAddress.toLowerCase()}"}) {
        reserveUSD
        dailyVolumeUSD
      }
    }
  `;
  const response = await fetch(graphUrl, {
    method: 'POST',
    body: JSON.stringify({ query }),
  });

  const sourceName = `${name} LP Fees`;
  const unknownValueSource = createValueSource(sourceName, uniformPerformance(0));

  if (!response.ok) {
    return unknownValueSource;
  }

  const pairDayResponse = await response.json();
  if (pairDayResponse.errors) {
    return unknownValueSource;
  }
  const pairDayData = pairDayResponse.data.pairDayDatas;

  let totalApy = 0;
  const performance: Performance = uniformPerformance(0);
  for (let i = 0; i < pairDayData.length; i++) {
    const volume = parseFloat(pairDayData[i].dailyVolumeUSD);
    const poolReserve = parseFloat(pairDayData[i].reserveUSD);
    const fees = volume * 0.003;
    totalApy += (fees / poolReserve) * 365 * 100;
    const currentApy = totalApy / (i + 1);
    if (i === 0) performance.oneDay = currentApy;
    if (i === 2) performance.threeDay = currentApy;
    if (i === 6) performance.sevenDay = currentApy;
    if (i === 29) performance.thirtyDay = currentApy;
  }

  return createValueSource(sourceName, performance);
}
