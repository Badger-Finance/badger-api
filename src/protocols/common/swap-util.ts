import { NotFound, UnprocessableEntity } from '@tsed/exceptions';
import { BigNumber, ethers } from 'ethers';
import { GraphQLClient } from 'graphql-request';
import { Chain } from '../../chains/config/chain.config';
import { Network } from '../../chains/enums/chain-network.enum';
import { uniV2LPAbi } from '../../config/abi';
import { SUSHISWAP_URL, UNISWAP_URL } from '../../config/constants';
import { getSdk as getUniswapSdk } from '../../graphql/generated/uniswap';
import { getTokenPriceData } from '../../prices/prices-util';
import { TokenPrice } from '../../tokens/interfaces/token-price.interface';

interface LiquidityData {
  contract: string;
  token0: string;
  token1: string;
  reserve0: number;
  reserve1: number;
  totalSupply: number;
}

interface GetReservesResponse {
  _reserve0: BigNumber;
  _reserve1: BigNumber;
  _blockTimestampLast: number;
}

export const getLiquidityData = async (chain: Chain, contract: string): Promise<LiquidityData> => {
  const pairContract = new ethers.Contract(contract, uniV2LPAbi, chain.provider);
  const totalSupply = parseFloat(ethers.utils.formatEther(await pairContract.totalSupply()));
  const token0 = await pairContract.token0();
  const token1 = await pairContract.token1();
  const reserves: GetReservesResponse = await pairContract.getReserves();
  const reserve0 = parseFloat(ethers.utils.formatEther(reserves._reserve0));
  const reserve1 = parseFloat(ethers.utils.formatEther(reserves._reserve1));
  return {
    contract: contract,
    token0: token0,
    token1: token1,
    reserve0: reserve0,
    reserve1: reserve1,
    totalSupply: totalSupply,
  };
};

export const getLiquidityPrice = async (graphUrl: string, contract: string): Promise<TokenPrice> => {
  const graphqlClient = new GraphQLClient(graphUrl);
  const graphqlSdk = getUniswapSdk(graphqlClient);
  const { pair } = await graphqlSdk.UniV2Pair({
    id: contract.toLowerCase(),
  });
  if (!pair) {
    throw new NotFound(`No pair found for ${contract}`);
  }
  if (pair.totalSupply === 0) {
    return {
      address: contract,
      usd: 0,
      eth: 0,
    };
  }
  const liquidityData: LiquidityData = {
    contract: contract,
    token0: pair.token0.id,
    token1: pair.token1.id,
    reserve0: pair.reserve0,
    reserve1: pair.reserve1,
    totalSupply: pair.totalSupply,
  };
  return resolveLiquidityPrice(liquidityData);
};

export const getOnChainLiquidityPrice = async (chain: Chain, contract: string): Promise<TokenPrice> => {
  try {
    const liquidityData = await getLiquidityData(chain, contract);
    if (liquidityData.totalSupply === 0) {
      return {
        address: contract,
        usd: 0,
        eth: 0,
      };
    }
    return resolveLiquidityPrice(liquidityData);
  } catch (err) {
    throw new NotFound(`No pair found for ${contract}`);
  }
};

const resolveLiquidityPrice = async (liquidityData: LiquidityData): Promise<TokenPrice> => {
  const { contract, token0, token1, reserve0, reserve1, totalSupply } = liquidityData;
  let t0Price = await getTokenPriceData(token0);
  let t1Price = await getTokenPriceData(token1);
  if (!t0Price && !t1Price) throw new UnprocessableEntity(`Token pair ${contract} cannot be priced`);
  if (!t0Price) {
    const t1Scalar = reserve0 / reserve1;
    t0Price = {
      address: token0,
      usd: t1Price.usd * t1Scalar,
      eth: t1Price.eth * t1Scalar,
    };
  }
  if (!t1Price) {
    const t0Scalar = reserve1 / reserve0;
    t1Price = {
      address: token1,
      usd: t0Price.usd * t0Scalar,
      eth: t0Price.eth * t0Scalar,
    };
  }
  const usdPrice = (t0Price.usd * reserve0 + t1Price.usd * reserve1) / totalSupply;
  const ethPrice = (t0Price.eth * reserve0 + t1Price.eth * reserve1) / totalSupply;
  return {
    address: contract,
    usd: usdPrice,
    eth: ethPrice,
  };
};

export const getSushiswapPrice = async (contract: string): Promise<TokenPrice> => {
  return getLiquidityPrice(SUSHISWAP_URL, contract);
};

export const getUniswapPrice = async (contract: string): Promise<TokenPrice> => {
  return getLiquidityPrice(UNISWAP_URL, contract);
};

export const getPancakeswapPrice = async (contract: string): Promise<TokenPrice> => {
  return getOnChainLiquidityPrice(Chain.getChain(Network.BinanceSmartChain), contract);
};
