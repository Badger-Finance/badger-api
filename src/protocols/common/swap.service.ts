import { Chain } from '../../chains/config/chain.config';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { TokenPrice } from '../../tokens/interfaces/token-price.interface';
import { ValueSource } from '../interfaces/value-source.interface';
import { getSwapValueSource } from './performance.utils';
import { getLiquidityPrice } from './swap.utils';

export abstract class SwapService {
  constructor(private graphUrl: string, private name: string) {}

  abstract getPairPerformance(
    chain: Chain,
    sett: SettDefinition,
    filterHarvestablePerformances?: boolean,
  ): Promise<ValueSource[]>;

  /**
   * Retrieve Uniswap v2 variant pool performance from trading fees.
   * @param poolAddress Liquidity pair contract address.
   * @param protocol Uniswap v2 variant type.
   */
  async getSwapPerformance(poolAddress: string): Promise<ValueSource> {
    return getSwapValueSource(this.graphUrl, this.name, poolAddress);
  }

  async getPairPrice(contract: string): Promise<TokenPrice> {
    return getLiquidityPrice(this.graphUrl, contract);
  }
}
