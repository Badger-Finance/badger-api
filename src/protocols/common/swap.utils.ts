import { Network } from '@badger-dao/sdk';
import { NotFound, UnprocessableEntity } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { GraphQLClient } from 'graphql-request';
import { Chain } from '../../chains/config/chain.config';
import { UNISWAP_URL } from '../../config/constants';
import { UniV2__factory } from '../../contracts';
import { getSdk as getUniswapSdk } from '../../graphql/generated/uniswap';
import { getPrice } from '../../prices/prices.utils';
import { TokenPrice } from '../../tokens/interfaces/token-price.interface';
import { formatBalance, getToken } from '../../tokens/tokens.utils';

interface LiquidityData {
  contract: string;
  token0: string;
  token1: string;
  reserve0: number;
  reserve1: number;
  totalSupply: number;
}

export async function getLiquidityData(chain: Chain, contract: string): Promise<LiquidityData> {
  const pairContract = UniV2__factory.connect(contract, chain.provider);
  const totalSupply = formatBalance(await pairContract.totalSupply());
  const token0 = await pairContract.token0();
  const token1 = await pairContract.token1();
  const token0Decimals = getToken(token0).decimals;
  const token1Decimals = getToken(token1).decimals;
  const reserves = await pairContract.getReserves();
  const reserve0 = formatBalance(reserves._reserve0, token0Decimals);
  const reserve1 = formatBalance(reserves._reserve1, token1Decimals);
  return {
    contract: contract,
    token0: token0,
    token1: token1,
    reserve0: reserve0,
    reserve1: reserve1,
    totalSupply: totalSupply,
  };
}

export const getLiquidityPrice = async (graphUrl: string, contract: string): Promise<TokenPrice> => {
  const graphqlClient = new GraphQLClient(graphUrl);
  const graphqlSdk = getUniswapSdk(graphqlClient);
  const { pair } = await graphqlSdk.UniV2Pair({
    id: contract.toLowerCase(),
  });
  if (!pair) {
    throw new NotFound(`No pair found for ${contract}`);
  }
  if (parseFloat(pair.totalSupply) === 0) {
    const token = getToken(contract);
    return {
      name: token.name,
      address: token.address,
      usd: 0,
      eth: 0,
    };
  }
  const liquidityData: LiquidityData = {
    contract: contract,
    token0: ethers.utils.getAddress(pair.token0.id),
    token1: ethers.utils.getAddress(pair.token1.id),
    reserve0: pair.reserve0,
    reserve1: pair.reserve1,
    totalSupply: pair.totalSupply,
  };
  return resolveLiquidityPrice(liquidityData);
};

export const getOnChainLiquidityPrice = async (chain: Chain, contract: string): Promise<TokenPrice> => {
  try {
    const liquidityData = await getLiquidityData(chain, contract);
    if (liquidityData.totalSupply === 0) {
      const token = getToken(contract);
      return {
        name: token.name,
        address: token.address,
        usd: 0,
        eth: 0,
      };
    }
    return resolveLiquidityPrice(liquidityData);
  } catch (err) {
    console.log(err);
    throw new NotFound(`No pair found for ${contract}`);
  }
};

const resolveLiquidityPrice = async (liquidityData: LiquidityData): Promise<TokenPrice> => {
  const { contract, token0, token1, reserve0, reserve1, totalSupply } = liquidityData;
  let t0Price = await getPrice(token0);
  let t1Price = await getPrice(token1);
  if (!t0Price && !t1Price) {
    throw new UnprocessableEntity(`Token pair ${contract} cannot be priced`);
  }
  if (!t0Price) {
    const t1Scalar = reserve0 / reserve1;
    const t0Info = getToken(token0);
    t0Price = {
      name: t0Info.name,
      address: t0Info.address,
      usd: t1Price.usd * t1Scalar,
      eth: t1Price.eth * t1Scalar,
    };
  }
  if (!t1Price) {
    const t0Scalar = reserve1 / reserve0;
    const t1Info = getToken(token1);
    t1Price = {
      name: t1Info.name,
      address: t1Info.address,
      usd: t0Price.usd * t0Scalar,
      eth: t0Price.eth * t0Scalar,
    };
  }
  const token = getToken(contract);
  const usdPrice = (t0Price.usd * reserve0 + t1Price.usd * reserve1) / totalSupply;
  const ethPrice = (t0Price.eth * reserve0 + t1Price.eth * reserve1) / totalSupply;
  return {
    name: token.name,
    address: token.address,
    usd: usdPrice,
    eth: ethPrice,
  };
};

export const resolveTokenPrice = async (chain: Chain, token: string, contract: string): Promise<TokenPrice> => {
  const { token0, token1, reserve0, reserve1 } = await getLiquidityData(chain, contract);
  const pricingToken = getToken(token);
  const isToken0 = pricingToken.address === token0;
  const knownToken = isToken0 ? token1 : token0;
  const [divisor, dividend] = isToken0 ? [reserve1, reserve0] : [reserve0, reserve1];
  const knownTokenPrice = await getPrice(knownToken);
  if (!knownTokenPrice) {
    throw new UnprocessableEntity(`Token ${pricingToken.name} cannot be priced`);
  }
  const scalar = divisor / dividend;
  const usdPrice = knownTokenPrice.usd * scalar;
  const ethPrice = knownTokenPrice.eth * scalar;
  return {
    name: pricingToken.name,
    address: pricingToken.address,
    usd: usdPrice,
    eth: ethPrice,
  };
};

export const getUniswapPrice = async (contract: string): Promise<TokenPrice> => {
  return getLiquidityPrice(UNISWAP_URL, contract);
};

// TODO: rethink how to generalize these fuctions, revisit strategy <> chain relations
export async function getPancakeswapPrice(contract: string): Promise<TokenPrice> {
  return getOnChainLiquidityPrice(Chain.getChain(Network.BinanceSmartChain), contract);
}

export async function getQuickswapPrice(contract: string): Promise<TokenPrice> {
  return getOnChainLiquidityPrice(Chain.getChain(Network.Polygon), contract);
}

export async function getArbitriumLiquidityPrice(contract: string): Promise<TokenPrice> {
  return getOnChainLiquidityPrice(Chain.getChain(Network.Arbitrum), contract);
}

export async function getFantomLiquidityPrice(contract: string): Promise<TokenPrice> {
  return getOnChainLiquidityPrice(Chain.getChain(Network.Fantom), contract);
}
