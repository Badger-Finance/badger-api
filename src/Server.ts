import { Configuration, Inject, PlatformApplication } from "@tsed/common";
import { controllers } from "./ControllerRegistry";
import methodOverride from "method-override";
import cookieParser from "cookie-parser";
import bodyParser from "body-parser";
import compress from "compression";

const rootDir = __dirname;

@Configuration({
  rootDir,
  acceptMimes: ["application/json"],
  mount: {
    "/v2/": controllers
  },
})
export class Server {
  @Inject()
  app!: PlatformApplication;

  /**
   * This method let you configure the express middleware required by your application to works.
   * @returns {Server}
   */
  public $beforeRoutesInit(): void | Promise<any> {
    this.app
      .use(cookieParser())
      .use(compress({}))
      .use(methodOverride())
      .use(bodyParser.json())
      .use(bodyParser.urlencoded({
        extended: true
      }));
  }
}
