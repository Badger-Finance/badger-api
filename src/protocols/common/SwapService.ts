import fetch from 'node-fetch';
import { Chain } from '../../chains/config/chain.config';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { TokenPrice } from '../../tokens/interfaces/token-price.interface';
import { Performance, uniformPerformance } from '../interfaces/performance.interface';
import { ValueSource } from '../interfaces/value-source.interface';
import { getLiquidityPrice } from './swap-util';

export abstract class SwapService {
  constructor(private graphUrl: string, private name: string) {}

  abstract getPairPerformance(
    chain: Chain,
    sett: SettDefinition,
    filterHarvestablePerformances?: boolean,
  ): Promise<ValueSource[]>;

  /**
   * Retrieve Uniswap v2 variant pool performance from trading fees.
   * @param poolAddress Liquidity pair contract address.
   * @param protocol Uniswap v2 variant type.
   */
  async getSwapPerformance(poolAddress: string): Promise<ValueSource> {
    // TODO: Move query to GraphService
    const query = `
      {
        pairDayDatas(first: 30, orderBy: date, orderDirection: desc, where:{pairAddress: "${poolAddress.toLowerCase()}"}) {
          reserveUSD
          dailyVolumeUSD
        }
      }
    `;
    const response = await fetch(this.graphUrl, {
      method: 'POST',
      body: JSON.stringify({ query }),
    });

    const unknownPerformance = uniformPerformance(0);
    const uknownValuSource = {
      name: `${this.name} LP Fees`,
      apy: unknownPerformance.threeDay,
      performance: unknownPerformance,
    };

    if (!response.ok) {
      return uknownValuSource;
    }

    const pairDayResponse = await response.json();
    if (pairDayResponse.errors) {
      return uknownValuSource;
    }
    const pairDayData = pairDayResponse.data.pairDayDatas;

    let totalApy = 0;
    const performance: Performance = uniformPerformance(0);
    for (let i = 0; i < pairDayData.length; i++) {
      const volume = parseFloat(pairDayData[i].dailyVolumeUSD);
      const poolReserve = parseFloat(pairDayData[i].reserveUSD);
      const fees = volume * 0.003;
      totalApy += (fees / poolReserve) * 365 * 100;
      const currentApy = totalApy / (i + 1);
      if (i === 0) performance.oneDay = currentApy;
      if (i === 2) performance.threeDay = currentApy;
      if (i === 6) performance.sevenDay = currentApy;
      if (i === 29) performance.thirtyDay = currentApy;
    }
    return {
      name: `${this.name} LP Fees`,
      apy: performance.threeDay,
      performance: performance,
    };
  }

  async getPairPrice(contract: string): Promise<TokenPrice> {
    return getLiquidityPrice(this.graphUrl, contract);
  }
}
