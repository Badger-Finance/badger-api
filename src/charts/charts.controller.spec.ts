// import { ChartTimeFrame } from '@badger-dao/sdk';
import { PlatformTest } from '@tsed/common';
import SuperTest from 'supertest';

import { NetworkStatus } from '../errors/enums/network-status.enum';
import { Server } from '../Server';
// import { TEST_ADDR } from '../test/constants';
// import { setupMockChain } from '../test/mocks.utils';
// import { setupVaultsHistoricDDB } from '../test/tests.utils';

describe('ChartsController', () => {
  let request: SuperTest.SuperTest<SuperTest.Test>;

  beforeAll(PlatformTest.bootstrap(Server));
  beforeAll(() => {
    request = SuperTest(PlatformTest.callback());
  });
  afterAll(PlatformTest.reset);

  describe('GET /v3/charts/vault', () => {
    describe('with a missing vault address', () => {
      it('returns 400, QueryParamError', async () => {
        const { body } = await request.get('/v3/charts/vault').expect(NetworkStatus.BadRequest);
        expect(body).toMatchSnapshot();
      });
    });

    // describe('get vault data with different timeframes', () => {
    //   it('should return vault data for YTD', async () => {
    //     setupVaultsHistoricDDB();

    //     const { body } = await request
    //       .get(`/v3/charts/vault?address=${TEST_ADDR}&timeframe=${ChartTimeFrame.YTD}`)
    //       .expect(NetworkStatus.Success);

    //     expect(body).toMatchSnapshot();
    //   });

    //   it('should return vault data for 1Y', async () => {
    //     setupVaultsHistoricDDB();

    //     const { body } = await request
    //       .get(`/v3/charts/vault?address=${TEST_ADDR}&timeframe=${ChartTimeFrame.Year}`)
    //       .expect(NetworkStatus.Success);

    //     expect(body).toMatchSnapshot();
    //   });
    // });
  });
});
