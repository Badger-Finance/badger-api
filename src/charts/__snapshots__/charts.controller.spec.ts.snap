// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartsController GET /v2/charts with an unsupported address returns a 404 1`] = `
Object {
  "message": "0x4206942069420694206942069420694206942069 is not a valid sett",
}
`;
